crates/sui-display/src/v2/lexer.rs

```diff
@@ -0,0 +1,947 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+#![allow(dead_code)]
+
+use std::fmt;
+
+/// Lexer for Display V2 format strings. Format strings are a mix of text and expressions.
+/// Expressions are enclosed in braces and may contain multiple alternates, separated by pipes, and
+/// each containing nested field, vector, or dynamic field accesses.
+#[derive(Debug)]
+pub(crate) struct Lexer<'s> {
+    /// Remaining input to be tokenized.
+    src: &'s str,
+
+    /// The number of bytes tokenized so far.
+    off: usize,
+
+    /// Whether the lexer is currently inside a text strand or an expression strand.
+    mode: Mode,
+}
+
+#[derive(Debug)]
+enum Mode {
+    Text,
+    Expr,
+}
+
+/// A lexeme is a token along with its offset in the source string, and the slice of source string
+/// that it originated from.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub(crate) struct Lexeme<'s>(pub Token, pub usize, pub &'s str);
+
+/// Like [Lexeme] but owns the slice of source string. Useful for capturing context in an error
+/// message.
+#[derive(Debug)]
+pub(crate) struct OwnedLexeme(Token, usize, String);
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub(crate) enum Token {
+    /// '@'
+    At,
+    /// ':'
+    Colon,
+    /// '::'
+    CColon,
+    /// ','
+    Comma,
+    /// '.'
+    Dot,
+    /// An identifier
+    Ident,
+    /// '<'
+    LAngle,
+    /// '{'
+    LBrace,
+    /// '['
+    LBracket,
+    /// '{{'
+    LLBrace,
+    /// '('
+    LParen,
+    /// A decimal number, optionally separated by underscores.
+    NumDec,
+    /// A hexadecimal number, prefixed with '0x' (not included in the span), optionally separated
+    /// by underscores.
+    NumHex,
+    /// '|'
+    Pipe,
+    /// '#'
+    Pound,
+    /// '>'
+    RAngle,
+    /// '}'
+    RBrace,
+    /// ']'
+    RBracket,
+    /// ')'
+    RParen,
+    /// '}}'
+    RRBrace,
+    /// Strings are surrounded by single quotes. Quotes and backslashes inside strings are escaped
+    /// with backslashes.
+    String,
+    /// A strand of text.
+    Text,
+
+    /// An unexpected byte in the input string.
+    Unexpected,
+
+    /// Whitespace around expressions.
+    Whitespace,
```

Comment:
Treating whitespace as an explicit token because there are some forms that are whitespace sensitive:

- Double square brackets.
- String literals with prefixes.
- Numeric literals with suffixes.